core/store/src/flat_state.rs

```diff
@@ -7,58 +7,89 @@
+    // TODO (#7327): store on top of final head (or earlier) so updates will
+    // only go forward.
+    #[derive(Clone)]
+    pub struct FlatState {
+        store: Store,
     }
 
-    #[allow(unused_variables)]
-    fn get_raw_ref(&self, key: &[u8]) -> Result<Option<Vec<u8>>, StorageError> {
-        #[cfg(feature = "protocol_feature_flat_state")]
-        return self
-            .store
-            .get(DBCol::FlatState, key)
-            .map_err(|_| StorageError::StorageInternalError);
-        #[cfg(not(feature = "protocol_feature_flat_state"))]
-        unreachable!();
-    }
+    impl FlatState {
+        fn get_raw_ref(&self, key: &[u8]) -> Result<Option<Vec<u8>>, StorageError> {
+            return self
+                .store
+                .get(crate::DBCol::FlatState, key)
+                .map_err(|_| StorageError::StorageInternalError);
+        }
 
-    /// Get value reference using raw trie key and state root. We assume that flat state contains data for this root.
-    /// To avoid duplication, we don't store values themselves in flat state, they are stored in `DBCol::State`. Also
-    /// the separation is done so we could charge users for the value length before loading the value.
-    /// TODO (#7327): support different roots (or block hashes).
-    pub fn get_ref(
-        &self,
-        _root: &CryptoHash,
-        key: &[u8],
-    ) -> Result<Option<ValueRef>, StorageError> {
-        match self.get_raw_ref(key)? {
-            Some(bytes) => {
-                ValueRef::decode(&bytes).map(Some).map_err(|_| StorageError::StorageInternalError)
+        /// Returns value reference using raw trie key and state root.
+        ///
+        /// We assume that flat state contains data for this root.  To avoid
+        /// duplication, we don't store values themselves in flat state, they
+        /// are stored in `DBCol::State`. Also the separation is done so we
+        /// could charge users for the value length before loading the value.
+        // TODO (#7327): support different roots (or block hashes).
+        pub fn get_ref(
+            &self,
+            _root: &CryptoHash,
+            key: &[u8],
+        ) -> Result<Option<ValueRef>, StorageError> {
+            match self.get_raw_ref(key)? {
+                Some(bytes) => ValueRef::decode(&bytes)
+                    .map(Some)
+                    .map_err(|_| StorageError::StorageInternalError),
+                None => Ok(None),
             }
-            None => Ok(None),
         }
     }
+
+    /// Possibly creates a new [`FlateState`] object backed by given storage.
+    ///
+    /// Always returns `None` if the `protocol_feature_flat_state` Cargo feature is
+    /// not enabled.  Otherwise, returns a new [`FlatState`] object backed by
+    /// specified storage if `use_flat_state` argument is true.
+    pub fn maybe_new(use_flat_state: bool, store: &Store) -> Option<FlatState> {
+        use_flat_state.then(|| FlatState { store: store.clone() })
```

Comment:
TBH, not a fan of `bool::then`

Reply:
I’m not a huge fan either to be honest, but it feels more readable here with one concise line.